### PR TITLE
Add context to documentable items & cleanup workspace

### PR DESCRIPTION
**Pull request by eureka-cpu**
This PR vastly increases readability of `forc doc`. It makes sure to evaluate all item context during the `descriptor` phase now, and remove all unnecessary newtype structs and enums. It also adds the basic item context for Sway `structs`, `storage`, `traits`, `abi` and `enums`.

There are some issues that I've stumbled upon during this, here are some screenshots:

1. Enum variant type spans are really just spans of the entire variant, not the type itself. #3634 
Screenshot from 2022-12-16 23-45-47
2. Methods on `abi` and `traits` also have this same issue, the span of the return type for functions that don't have a return type I assumed would be `()` (unit type) but as you can see it's just giving me the span of the entire function signature. #3635 
Screenshot from 2022-12-16 23-46-04
3. Less important but just an eyes sore, normal comments get removed during formatting but it seems that doc comments do not which I guess makes sense because they are really attributes under the hood. This problem should go away once I add links to everything since I will have to change how we go about formatting.
Screenshot from 2022-12-16 23-46-22
Screenshot from 2022-12-16 23-45-57

### CODE REVIEW

File: sway-core/src/language/ty/declaration/declaration.rs [hunk 1]
```diff
@@ -468,6 +468,22 @@ impl TyDeclaration {
...
+        match self {
+            StructDeclaration(_) => "struct",
+            EnumDeclaration(_) => "enum",
+            TraitDeclaration(_) => "trait",
+            AbiDeclaration(_) => "abi",
+            StorageDeclaration(_) => "storage",
+            ImplTrait(_) => "impl_trait",
+            FunctionDeclaration(_) => "fn",
+            ConstantDeclaration(_) => "constant",
+            _ => "", // non-documentable items, this arm is unreachable
```
Comment (Dentosal) [inline sway-core/src/language/ty/declaration/declaration.rs:483]: If the arm is really unreachable, using `unreachable!("...")` would be nicer than just a comment

Comment (eureka-cpu) [reply sway-core/src/language/ty/declaration/declaration.rs:483]: Okay nice! Thanks for the review

**Review by mitchmindtree — DISMISSED** [review]
All looking good! I've left some optional suggestions, but otherwise utACK :+1:

File: forc-plugins/forc-doc/src/main.rs [hunk 2]
```diff
@@ -58,63 +58,71 @@ pub fn main() -> Result<()> {
...
-    let compilation = pkg::check(&plan, silent, &type_engine)?
-        .pop()
-        .expect("there is guaranteed to be at least one elem in the vector");
-    let raw_docs: Documentation =
-        Document::from_ty_program(&compilation, no_deps, document_private_items)?;
-    // render docs to HTML
-    let rendered_docs: RenderedDocumentation =
-        RenderedDocument::from_raw_docs(&raw_docs, project_name);
+    if let Some(compilation) = pkg::check(&plan, silent, &type_engine)?.pop() {
+        if let Some((_, Some(typed_program))) = compilation.value {
```
[the remark this answers is not in the document]
Comment (eureka-cpu) [reply forc-plugins/forc-doc/src/main.rs:62]: Ah yeah that looks a lot better!

File: forc-plugins/forc-doc/src/descriptor.rs [hunk 3]
```diff
@@ -1,140 +1,312 @@
...
                 if !document_private_items && struct_decl.visibility.is_private() {
                     Ok(Descriptor::NonDocumentable)
                 } else {
-                    Ok(Descriptor::Documentable {
+                    let item_name = struct_decl.name.as_str().to_string();
+                    let attrs_opt = if !struct_decl.attributes.is_empty() {
+                        Some(attrsmap_to_html_string(&struct_decl.attributes))
+                    } else {
+                        None
+                    };
```
[the remark this answers is not in the document]
Comment (eureka-cpu) [reply forc-plugins/forc-doc/src/descriptor.rs:42]: Ah this looks a lot better, thanks!

**Comment by eureka-cpu** [comment]
> @eureka-cpu I tried checking out your branch and running `forc doc --open` in the standard library but got the following error:
> 
> ```
> thread 'main' panicked at 'internal error: entered unreachable code: these items are non-documentable', forc-plugins/forc-doc/src/render.rs:524:22
> ```
> 
> On `master`, I do not get this error and I'm able to generate the docs.

ah i know what this is, I added the `unreachable` macro to a section that includes `impl_trait`s.

**Review by mitchmindtree — APPROVED** [review]

**Review by JoshuaBatty — APPROVED** [review]

**Pull request merged**